lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -35,136 +32,369 @@
+          nlaBuilder.getStringAttr(circtNamespace.newName("memNLA")),
+          nlaBuilder.getArrayAttr(namepath));
 
+      // Create the path operation.
+      memoryHierPaths.emplace_back(createPathRef(finalInst, nla, builderOM));
+    }
+    auto hierpaths = builderOM.create<ListCreateOp>(
+        ListType::get(context, PathType::get(context).cast<PropertyType>()),
+        memoryHierPaths);
     SmallVector<Value> memFields;
-    for (auto field : memoryParamNames)
-      memFields.push_back(createConstField(
-          llvm::StringSwitch<TypedAttr>(field)
-              .Case("name", om::SymbolRefAttr::get(mem))
+
+    auto object = builderOM.create<ObjectOp>(memorySchemaClass, mem.getName());
+    SmallVector<Value> extraPortsList;
+    ClassType extraPortsType;
+    for (auto attr : mem.getExtraPortsAttr()) {
+
+      auto port = cast<DictionaryAttr>(attr);
+      auto portName = createConstField(port.getAs<StringAttr>("name"));
+      auto direction = createConstField(port.getAs<StringAttr>("direction"));
+      auto width = createConstField(port.getAs<IntegerAttr>("width"));
+      auto extraPortsObj =
+          builderOM.create<ObjectOp>(extraPortsClass, "extraPorts");
+      extraPortsType = extraPortsObj.getType();
+      auto inPort = builderOM.create<ObjectSubfieldOp>(extraPortsObj, 0);
+      builderOM.create<PropAssignOp>(inPort, portName);
+      inPort = builderOM.create<ObjectSubfieldOp>(extraPortsObj, 2);
+      builderOM.create<PropAssignOp>(inPort, direction);
+      inPort = builderOM.create<ObjectSubfieldOp>(extraPortsObj, 4);
+      builderOM.create<PropAssignOp>(inPort, width);
+      extraPortsList.push_back(extraPortsObj);
+    }
+    auto extraPorts = builderOM.create<ListCreateOp>(
+        memorySchemaClass.getPortType(22), extraPortsList);
+    for (auto field : llvm::enumerate(memoryParamNames)) {
+      auto propVal = createConstField(
+          llvm::StringSwitch<TypedAttr>(field.value())
+              .Case("name", builderOM.getStringAttr(mem.getName()))
               .Case("depth", mem.getDepthAttr())
               .Case("width", mem.getDataWidthAttr())
               .Case("maskBits", mem.getMaskBitsAttr())
               .Case("readPorts", mem.getNumReadPortsAttr())
               .Case("writePorts", mem.getNumWritePortsAttr())
               .Case("readwritePorts", mem.getNumReadWritePortsAttr())
               .Case("readLatency", mem.getReadLatencyAttr())
-              .Case("writeLatency", mem.getWriteLatencyAttr())));
+              .Case("writeLatency", mem.getWriteLatencyAttr())
+              .Case("hierarchy", {})
+              .Case("inDut", BoolAttr::get(context, inDut))
+              .Case("extraPorts", {}));
+      if (!propVal) {
+        if (field.value().equals("hierarchy"))
+          propVal = hierpaths;
+        else
+          propVal = extraPorts;
+      }
+
+      auto inPort =
+          builderOM.create<ObjectSubfieldOp>(object, 2 * field.index());
```

Comment:
Can you add a comment about the 2? I know it's because we're making a "simple class" with input tied to output for each field, but it will be helpful to know what's going on here in the future.

Reply:
Added the comment explaining the logic.